{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       Symbol Protein Halflife\n",
      "ID                            \n",
      "MG_001   dnaN             2,40\n",
      "MG_002    NaN             2,50\n",
      "MG_003   gyrB             2,50\n",
      "MG_004   gyrA             2,60\n",
      "MG_005   serS             2,60\n",
      "MG_006    tmk             3,90\n",
      "MG_007   holB             2,90\n",
      "MG_008   mnmE             4,10\n",
      "MG_009   tatD             4,80\n",
      "MG_010   dnaE             4,43\n",
      "MG_011    NaN             4,43\n",
      "MG_012   rimK             4,40\n",
      "MG_013   folD             3,20\n",
      "MG_014   mdlB             3,70\n",
      "MG_015   mdlB             5,60\n",
      "MG_018  hepA2             4,43\n",
      "MG_019   dnaJ             2,50\n",
      "MG_020   pepP             4,43\n",
      "MG_021   metG             2,10\n",
      "MG_022   rpoE             4,43\n",
      "MG_023    fba             7,20\n",
      "MG_024   ychF             2,70\n",
      "MG_025   trsB             4,43\n",
      "MG_026    efp             3,20\n",
      "MG_027   nusB             3,30\n",
      "MG_028    NaN             4,43\n",
      "MG_029    NaN             4,43\n",
      "MG_030    upp             4,10\n",
      "MG_031   polC             4,43\n",
      "MG_032    NaN             4,43\n",
      "...       ...              ...\n",
      "MG_442   rbgA             4,43\n",
      "MG_443    NaN             4,43\n",
      "MG_444   rplS             6,40\n",
      "MG_445   trmD             6,10\n",
      "MG_446   rpsP             5,80\n",
      "MG_447   norM             4,43\n",
      "MG_448   msrB             6,20\n",
      "MG_449    NaN              NaN\n",
      "MG_450    NaN             4,43\n",
      "MG_451    tuf             7,00\n",
      "MG_452    NaN             4,43\n",
      "MG_453   galU             4,90\n",
      "MG_454    ohr             7,50\n",
      "MG_455   tyrS             2,00\n",
      "MG_456    NaN             4,43\n",
      "MG_457   ftsH             2,20\n",
      "MG_458    hpt             6,30\n",
      "MG_459    NaN             4,43\n",
      "MG_460    ldh            10,50\n",
      "MG_461    NaN             4,43\n",
      "MG_462   gltX             2,80\n",
      "MG_463   ksgA             2,20\n",
      "MG_464   yidC             4,43\n",
      "MG_465   rnpA             3,30\n",
      "MG_466  rpL34             4,50\n",
      "MG_467    NaN             4,43\n",
      "MG_468    NaN             4,43\n",
      "MG_526    NaN             4,43\n",
      "MG_469   dnaA             1,90\n",
      "MG_470    soj             4,43\n",
      "\n",
      "[482 rows x 2 columns]\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import sqlite3 as lite\n",
    "import cPickle as pickle\n",
    "\n",
    "\n",
    "####INput aus Excel-Tabellen\n",
    "#Einlesen der Gen-, Metabolit- und Mediumsspezifikation\n",
    "genes = pd.read_csv('genes.csv',header=4, index_col=1)\n",
    "metabolites =pd.read_csv('metabolite.csv',header=4, index_col=0, na_values=['NaN'], keep_default_na=False)\n",
    "media = pd.read_csv('media.csv',header=3,index_col=0, na_values=['NaN'],keep_default_na=False)\n",
    "mrna = pd.read_csv('mRNA.csv',header=4, index_col=0 , na_values=[' '],keep_default_na=False)\n",
    "protein_hl = protein_hl['M9 Manually Imputed']\n",
    "protein_hl.columns =['protein half-life']\n",
    "\n",
    "#Editieren der Dataframes\n",
    "#Herauswerfen aller unwichtigen Spalten\n",
    "metabolites = metabolites['Name']\n",
    "genes = genes[['SwissProt','GenBank','Name','Sequence']]\n",
    "media = media['Concentration mM']\n",
    "mrna = mrna[['Symbol', 'Protein Halflife']]\n",
    "\n",
    "print mrna\n",
    "#Konkatenieren der Listen an entsprechender Stelle und Herauswerfen der NaN Werte\n",
    "conc = pd.concat([metabolites, media], axis=1)\n",
    "conc = conc[pd.notnull(conc['Concentration mM'])]\n",
    "\n",
    "\n",
    "##Laden des vollstaendigen Genoms in einen Objekt\n",
    "with open('sequence.fasta','r') as f:\n",
    "\tgenome = f.readlines()\n",
    "f.close()\n",
    "#Herauswerfen der ersten Zeile und zusammenhaengen der Sequenzen in einen String\n",
    "genome = [line[:-1] for line in genome if line[0] != '>']\n",
    "genome = ''.join(genome)\n",
    "\n",
    "\n",
    "#name=genes/metabolites/media\n",
    "def writeData(data,name):\n",
    "\tf = open(name,'w')\n",
    "\twith f:\n",
    "\t\tf.write(data)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
